scripts/prepare_pubmedqa_instruction.py

```diff
@@ -0,0 +1,156 @@
+"""Implementation derived from https://github.com/tloen/alpaca-lora"""
+import json
+import os
+import sys
+from pathlib import Path
+from typing import List,Optional
+
+import requests
+import torch
+from torch.utils.data import random_split
+from tqdm import tqdm
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_gpt.tokenizer import Tokenizer
+
+
+def prepare(
+    destination_path: Path = Path("data/pubmed-qa-instruct"),
+    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    mask_inputs: bool = False,
+    data_file_name: str = "pubmed_qa_instruction_cleaned.json",
+    data_repo_id: str = "FedML/PubMedQA_instruction",
+    ignore_index: int = -1,
+    access_token: Optional[str] = os.getenv("HF_TOKEN"),
+    max_seq_length: Optional[int] = None,
+) -> None:
+    """Prepare the pubmedQA Instruction dataset for instruction tuning.
+
+    The output is a training and test dataset saved as `train.pt` and `test.pt`,
+    which stores the preprocessed and tokenized prompts and labels.
+    """
+
+    if access_token is None:
+        raise ValueError(
+            "pubmedQA Instruction requires authentication, please set the `HF_TOKEN=your_token` environment"
+            " variable or pass --access_token=your_token. You can find your token by visiting"
+            " https://huggingface.co/settings/tokens"
+        )
+    
+    if max_seq_length is None:
+        with open(checkpoint_dir / "lit_config.json", "r", encoding="utf-8") as file:
+            config = json.load(file)
+            max_seq_length = config["block_size"]
+
+    destination_path.mkdir(parents=True, exist_ok=True)
+    data_file_path = destination_path / data_file_name
+    print("Loading data file...")
+
+    from datasets import load_dataset
+
+    dataset = load_dataset(data_repo_id, use_auth_token=access_token)
+    train_set = dataset["train"]
+    test_set = dataset["test"]
+    
+    # download_if_missing(data_file_path, data_file_url)
+
+    # with open(data_file_path, "r", encoding="utf-8") as file:
+    #     data = file.readlines()
+    #     data = [json.loads(line) for line in data]
+    # for item in data:
+    #     item["input"] = item.pop("context")
+    #     item["output"] = item.pop("response")
+
+    print("Loading tokenizer...")
+    tokenizer = Tokenizer(checkpoint_dir)
+
+    train_set, test_set = list(train_set), list(test_set)
+
+    print(f"train has {len(train_set):,} samples")
+    print(f"test has {len(test_set):,} samples")
+
+    print("Processing train split ...")
```

Comment:
Although I'm not deeply familiar with the HF apis yet, seems like this processing could be done in terms of a `dataset.map`, which may be a bit quicker if this ends up taking a while.

Reply:
Not sure about the performance differences between dataset.map and this but it's something to keep in mind for larger datasets. This dataset is small enough where the difference probably won't matter, especially since we won't run this too frequently.